From: Carlos Garnacho <carlosg@gnome.org>
Date: Fri, 5 Jun 2020 22:39:55 +0200
Subject: clutter: Add ClutterSeat::query_state() vmethod

This will query a device state; position and modifiers.
Note that ClutterSeat subclasses don't implement the vmethod yet,
so calling clutter_seat_query_state() may crash ATM.

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1403>
---
 clutter/clutter/clutter-seat.c | 17 +++++++++++++++++
 clutter/clutter/clutter-seat.h | 13 +++++++++++++
 2 files changed, 30 insertions(+)

diff --git a/clutter/clutter/clutter-seat.c b/clutter/clutter/clutter-seat.c
index 13a862b..3254701 100644
--- a/clutter/clutter/clutter-seat.c
+++ b/clutter/clutter/clutter-seat.c
@@ -749,3 +749,20 @@ clutter_seat_get_touch_mode (ClutterSeat *seat)
 
   return touch_mode;
 }
+
+gboolean
+clutter_seat_query_state (ClutterSeat          *seat,
+                          ClutterInputDevice   *device,
+                          ClutterEventSequence *sequence,
+                          graphene_point_t     *coords,
+                          ClutterModifierType  *modifiers)
+{
+  g_return_val_if_fail (CLUTTER_IS_SEAT (seat), FALSE);
+  g_return_val_if_fail (CLUTTER_IS_INPUT_DEVICE (device), FALSE);
+
+  return CLUTTER_SEAT_GET_CLASS (seat)->query_state (seat,
+                                                     device,
+                                                     sequence,
+                                                     coords,
+                                                     modifiers);
+}
diff --git a/clutter/clutter/clutter-seat.h b/clutter/clutter/clutter-seat.h
index aa88d73..082c3e9 100644
--- a/clutter/clutter/clutter-seat.h
+++ b/clutter/clutter/clutter-seat.h
@@ -113,6 +113,12 @@ struct _ClutterSeatClass
                          int          x,
                          int          y);
 
+  gboolean (* query_state) (ClutterSeat          *seat,
+                            ClutterInputDevice   *device,
+                            ClutterEventSequence *sequence,
+                            graphene_point_t     *coords,
+                            ClutterModifierType  *modifiers);
+
   /* Event platform data */
   void (* copy_event_data) (ClutterSeat        *seat,
                             const ClutterEvent *src,
@@ -194,4 +200,11 @@ void clutter_seat_warp_pointer (ClutterSeat *seat,
 CLUTTER_EXPORT
 gboolean clutter_seat_get_touch_mode (ClutterSeat *seat);
 
+CLUTTER_EXPORT
+gboolean clutter_seat_query_state (ClutterSeat          *seat,
+                                   ClutterInputDevice   *device,
+                                   ClutterEventSequence *sequence,
+                                   graphene_point_t     *coords,
+                                   ClutterModifierType  *modifiers);
+
 #endif /* CLUTTER_SEAT_H */
